### PR TITLE
feat(trainer): add enable_benchmarking option

### LINKED ISSUE #370: Support to turn on cudnn benchmark mode
Cudnn benchmark mode is [recommended for improved performance](https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936/11).

It would be great to have a one-liner option to turn this on. This could be a `Trainer` constructor arg. I'm happy to do a quick PR/documentation update for this, just want to get feedback/confirmation before I do it.

Only downside/complexity is [when determinism is also enabled](https://github.com/pytorch/pytorch/issues/6351). AFAICT, there's no option for determinism in Lightning, so seems like this shouldn't be a problem.

### PR DESCRIPTION
**Pull request by Ir1d**
closes #370
I don't think we need tests for this option for now. Tests can be added when this flag includes more options.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes #370 (issue).

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -483,12 +484,23 @@ def __init__(
                     profiler = AdvancedProfiler()
                     trainer = Trainer(profiler=profiler)
 
+            enable_benchmarking (bool): If true enables benchmarking.
+                Example::
+
+                    # default disabled by the Trainer
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: I do not think this need example like this...

Comment (Ir1d) [reply pytorch_lightning/trainer/trainer.py]: haha i followed the style of previous lines. let me remove this

Comment (Ir1d) [reply pytorch_lightning/trainer/trainer.py]: wait a sec, the other options are enabled by default right? and this option i think should be disabled by default

**Comment by Borda** [comment]
@Ir1d I wee that Will made some changes, could you check it and push so it triggers the new CI

**Comment by Ir1d** [comment]
@Borda you can re-trigger the build on travis page since you are a member of the org. I don't have other things to push in this PR

**Comment by Borda** [comment]
well the point is that we moved to GitHub CI and that need to add a commit to update all these things lol

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -89,7 +89,8 @@ def __init__(
             num_sanity_val_steps=5,
             truncated_bptt_steps=None,
             resume_from_checkpoint=None,
-            profiler=None
+            profiler=None,
+            benchmark=False
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: add "," so wth adding new param it makes just one line :]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: my point was to write a sentence what the default value is, do not pass redundant example...

Comment (Ir1d) [reply pytorch_lightning/trainer/trainer.py]: well I think `default disabled by the Trainer` is explicit enough.. I made a little change. please take a look :smile:

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -483,12 +484,24 @@ def __init__(
                     profiler = AdvancedProfiler()
                     trainer = Trainer(profiler=profiler)
 
+            benchmark (bool): If true enables cudnn.benchmark. From PyTorchdocs: "This flag allows you to enable the inbuilt cudnn auto-tuner
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: you may rather use citations, see https://docutils.sourceforge.io/docs/user/rst/quickref.html#citations

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -483,12 +484,24 @@ def __init__(
                     profiler = AdvancedProfiler()
                     trainer = Trainer(profiler=profiler)
 
+            benchmark (bool): If true enables cudnn.benchmark. From PyTorchdocs: "This flag allows you to enable the inbuilt cudnn auto-tuner 
+                to find the best algorithm to use for your hardware."
+                Example::
+
+                    # default disabled(false) by the Trainer
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: drop this example

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (Ir1d) [reply pytorch_lightning/trainer/trainer.py]: will brought this line in..
I'm not on the side of my computer at this moment, so I can't bring a fix.

**Comment by Ir1d** [comment]
The checks are passed and now ready to land.

**Comment by williamFalcon** [comment]
@Ir1d this needs a test. Just run a simple model for a few steps with this flag enabled.

Tests are not just for functionality but also for test coverage.

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: the style was pruned recently and all examples saying this is default value was removed... ^^

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: @Ir1d also here ^^

**Comment by williamFalcon** [comment]
@Ir1d mind doing a rebase?

**Comment by Borda** [comment]
@williamFalcon ^^

**Comment by awaelchli** [comment]
@Ir1d What's the purpose of the test you wrote. It has nothing to do with the benchmarking option and just tests if the trainer completes.

**Comment by Ir1d** [comment]
@awaelchli Feel free to enlighten me on how to test this benchmarking option. The tests are added for test coverage, as suggested above by william.

**Comment by awaelchli** [comment]
@Ir1d Ok, I see. Maybe just assert the value of torch.backends.cudnn.benchmark before and after fit? We would expect it to never be turned off during fit.

**Comment by Ir1d** [comment]
Thank you for your suggestions @awaelchli , I've added the assert statements.

**Comment by Borda** [comment]
@Ir1d @awaelchli I love this spirit! :zap:

**Comment by awaelchli** [comment]
nice, great addition! GPU tests pass on my machine.

**Comment by awaelchli** [comment]
I think @Borda made changes in #936 and this pr was not rebased after that, so it broke tests.

**Comment by awaelchli** [comment]
nvm, I think merge of #889 fixed it.

**Pull request merged**